src/System.Linq/src/System/Linq/Partition.cs

```diff
@@ -42,22 +42,18 @@ internal interface IPartition<TElement> : IIListProvider<TElement>
 
         IPartition<TElement> Take(int count);
 
-        TElement ElementAt(int index);
+        TElement TryGetElementAt(int index, out bool found);
 
-        TElement ElementAtOrDefault(int index);
+        TElement TryGetFirst(out bool found);
 
-        TElement First();
-
-        TElement FirstOrDefault();
-
-        TElement Last();
-
-        TElement LastOrDefault();
+        TElement TryGetLast(out bool found);
```

Comment:
Since it complicates the code, we should really understand whether it has the intended perf benefit; if it doesn't, I'd prefer the code remain simpler and more consistent with the standard ```bool Try*(out T)``` pattern.  If it has the desired impact, a comment to explain it would be good (so that someone in the future doesn't say "well this is really strange, we should change it to match the standard pattern").